File: android/app/src/main/java/org/fs/news/model/intent/LoadMoreNewsIntent.kt
/*
 *  Copyright (C) 2019 Fatih, News Stand Android Kotlin.
 *  
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.fs.news.model.intent

import io.reactivex.Observable
import io.reactivex.schedulers.Schedulers
import org.fs.architecture.mvi.common.*
import org.fs.news.model.NewsModel
import org.fs.news.model.entity.News
import org.fs.news.net.model.Resource
import org.fs.news.net.repo.NewsRepository
import org.fs.news.util.Operations.Companion.LOAD_MORE
import java.io.IOException

class LoadMoreNewsIntent(
  private val source: String,
  private val page: Int,
  private val newsRepository: NewsRepository): ObservableIntent<NewsModel>() {

  override fun invoke(): Observable<Reducer<NewsModel>> = newsRepository.news(source, page)
    .concatMap(::success)
    .onErrorResumeNext(::failure)
    .startWith(initial())
    .subscribeOn(Schedulers.io())

  private fun success(resource: Resource<List<News>>): Observable<Reducer<NewsModel>> = when(resource) {
    is Resource.Success<List<News>> -> Observable.just(
      { o -> o.copy(state = Operation(LOAD_MORE), data = resource.data ?: emptyList(), totalSize = resource.totalResult ?: 0) },
      { o -> o.copy(state = Idle, data = emptyList(), totalSize = 0) })
    is Resource.Failure<List<News>> -> Observable.just(
      { o -> o.copy(state = Failure(IOException(resource.message))) },
      { o -> o.copy(state = Idle) })
  }

  private fun failure(error: Throwable): Observable<Reducer<NewsModel>> = Observable.just(
    { o -> o.copy(state = Failure(error)) },
    { o -> o.copy(state = Idle) })

  private fun initial(): Reducer<NewsModel> = { o -> o.copy(state = Operation(LOAD_MORE), data = emptyList(), totalSize = 0) }
}

File: ios/NewsStand/Model/Intent/LoadMoreNewsIntent.swift
//
//  LoadMoreNewsIntent.swift
//  NewsStand
//
//  Created by Fatih Şen on 29.06.2019.
//  Copyright © 2019 Fatih Şen. All rights reserved.
//

import Foundation
import RxSwift
import MVICocoa

class LoadMoreNewsIntent: ObservableIntent<NewsModel> {
	
	private let newsRepository: NewsRepository
	private let source: String
	private let page: Int
	
	init(source: String, page: Int, newsRepository: NewsRepository) {
		self.newsRepository = newsRepository
		self.page = page
		self.source = source
	}
	
	override func invoke() -> Observable<Reducer<NewsModel>> {
		return newsRepository.news(source: source, page: page)
			.concatMap(success(_ :))
			.catchError(failure(_:))
			.startWith(initial())
			.subscribeOn(MainScheduler.asyncInstance)
	}
	
	private func initial() -> Reducer<NewsModel> {
		return { o in o.copy(state: .operationOf(loadMore), data: [], totalResult: 0) }
	}
	
	private func success(_ resource: Resource<[News]>) -> Observable<Reducer<NewsModel>> {
		switch resource {
		case .success(let data, let totalResult): return Observable.of(
		{ o in o.copy(state: .operationOf(loadMore), data: data ?? [], totalResult: totalResult ?? 0)},
		{ o in o.copy(state: .idle, data: [], totalResult: 0) })
		case .failure(_, let message): return Observable.of(
		{ o in o.copy(state: .failure(MVIError.of(description: message ?? ""))) },
		{ o in o.copy(state: .idle) })
		}
	}
	
	private func failure(_ error: Error) -> Observable<Reducer<NewsModel>> {
		return Observable.of(
			{ o in o.copy(state: .failure(error)) },
			{ o in o.copy(state: .idle) })
	}
}
